docs/tutorials/forum-application/7.vote-extension.md

```diff
@@ -0,0 +1,125 @@
+
+This allows validators to have access to all vote extensions at the next height. They can then use the data in the
+`vote extensions` to inform the transactions that make it into the next block.
+
+**Following is the code for the `ExtendVote` function:**
+
+```go
+// ExtendVote returns curse words as vote extensions
+func (app *ForumApp) ExtendVote(_ context.Context, _ *abci.ExtendVoteRequest) (*abci.ExtendVoteResponse, error) {
+	fmt.Println("Executing Application ExtendVote")
+
+	return &abci.ExtendVoteResponse{VoteExtension: []byte(app.CurseWords)}, nil
+}
+```
+
+**Explanation of code:**
+
+`ExtendVote` function takes two parameters: a `context.Context` and a pointer to an abci.`ExtendVoteRequest` struct.
+It returns a pointer to an `ExtendVoteResponse` struct and an error.
+
+The method implementation simply returns a new instance of `ExtendVoteResponse` with the `VoteExtension` field set to
+the value of `app.CurseWords.` The `app.CurseWords` is expected to be a byte array containing the vote extension data.
+
+The `ExtendVoteResponse` struct is used to encapsulate the response data for the `ExtendVote` method. By setting
+the `VoteExtension` field, the method includes the application-specific vote extension data in the response.
+
+In this implementation, the `ExtendVote` method in the ForumApp application returns the application-specific `vote extension`
+data stored in the `app.CurseWords` variable.
+
+*Tip: The `vote extensions` are opaque to the consensus algorithm but visible to the application, allowing for a variety
+of use cases like **price oracles, encrypted mempools, and threshold cryptography.***
+
+## VerifyVoteExtension
+
+The `VerifyVoteExtension` method allows applications to verify the `VoteExtension` data attached to each `pre-commit` message.
+
+When a validator is preparing to send a `pre-commit` vote, it first calls `ExtendVote` to generate a `VoteExtension.`
+This `VoteExtension` is broadcast along with the `pre-commit` vote.
+
+Other validators also call `ExtendVote` to generate their own `vote extensions.` However, not all validators will generate
+the same `vote extension.`
+
+When a validator receives a `pre-commit` vote with an attached `vote extension`, it calls `VerifyVoteExtension` to
+validate the `vote extension.`
+
+If the vote extension is successfully verified, the `pre-commit` vote is included in the tally. If validation fails,
+the entire `pre-commit` message is ignored.
+
+**Following is the blurb of code for the `VerifyVoteExtension` function:**
+
+```go
+// VerifyVoteExtension verifies the vote extensions and ensure they include the curse words
+// It will not be called for extensions generated by this validator
+func (app *ForumApp) VerifyVoteExtension(_ context.Context, req *abci.VerifyVoteExtensionRequest) (*abci.VerifyVoteExtensionResponse, error) {
+	fmt.Println("Executing Application VerifyVoteExtension")
+
+	if _, ok := app.valAddrToPubKeyMap[string(req.ValidatorAddress)]; !ok {
```

Comment:
How do you manage `valAddrToPubKeyMap`?

Reply:
in the `app.go`, in the `InitChain` method they get populated

```
	for _, v := range req.Validators {
		err := app.updateValidator(v)
		if err != nil {
			return nil, err
		}
	}
```

but other than that I don't think there's logic to update this map outside of the `InitChain`

Reply:
so the application only allow updates to validator from genesis, I think it is beyond the scope for this tutorial app to handle validator updates since it would require quite some additional logic in order to support that. So I'm assuming it's OK to leave the way it is.

Reply:
Yeah, no val updates needed for the tutorial. But what happens if the node crashes? Do we recover this info?